### PR TITLE
Orders list with last update

### LINKED ISSUE #11796: Show the last updated timestamp on the orders list
# Task

- [ ] Update UI to a timestamp of when the data was last updated.

# Notes

- The timestamp should also reflect the time when the orders where updated in the foreground (pull to refresh, etc)

## TODO

- [ ] Define UI Mock up

### LINKED ISSUE #11799: Add Updating indicator to the orders list
# Task

- [x] Update UI to show an updating indicator when the orders list is updating its data.

## TODO

- [ ] Define UI Mock up

### PR DESCRIPTION
**Pull request by atorresveiga**
⚠️ This PR depends on this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3056)

Closes: #11796
Closes: #11799

### Why
This PR shows the latest order sync time in the orders list screen and refresh the first page only when the data is outdated.

### Description
This pull request introduces a mechanism to prevent automatic order list syncing if less than 30 minutes have passed since the last successful sync. This optimization aims to:

- Reduce unnecessary API calls: By preventing frequent syncs, we can minimize API load and potentially improve overall performance.
- Optimize battery life: Fewer background operations can contribute to extended battery life on user's devices.

Implementation Details:

- The last order list sync timestamp will be stored in device preferences.
- Before initiating a new automatic sync, the code will check the time difference between the current time and the stored timestamp.
- If the difference is less than 30 minutes, the automatic sync will be skipped.

Benefits:

- Improved application performance
- Reduced API calls
- Potentially extended battery life

Additional Notes: 

- Users can still manually trigger a sync even if 30 minutes haven't passed by using the pull-to-refresh gesture.

### Testing

TC1 - Opening the order list for the first time
1. Open the app
2. Navigate to the orders section
3. Check that the orders list is displayed successfully
4. Check that the last update section is update with the current time

TC2 - Opening the order list with cached data
1. Run TC1
2. Close the app and wait some minutes
3. Open the app
4. Navigate to the orders section
5. Check that the last update section is displayed with the timestamp from TC1
6. Check that the data is displayed from the DB and no additional request are made to the API to display the first page
7. Pull to refresh
8. Check that the app refresh data from the API
9. Check that the last update section is updated 

TC3 - Filters also works with cached data
1. Open the app
2. Navigate to the orders section
3. Check that the orders list is displayed successfully
4. Tap on Filters
5. Set a filter and press Show Orders
6. Check that the expected orders are displayed
7. Go back to Filters and clear all the filters
8. Check that the orders list data is displayed from the DB without requests to the API 
9. Go back to Filters
10. Set again the filter from step 5
11. Check that the orders list data is displayed from the DB without requests to the API 

TC4
1. Check that searching orders keeps working as expected 

### Images/gif

https://github.com/user-attachments/assets/ce9c4c4a-a838-4a39-8b8e-9da157b64e89

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>UpdateData</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/227">19.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/FetchOrdersRepository.kt [hunk 1]
```diff
@@ -60,4 +67,16 @@ class FetchOrdersRepository @Inject constructor(private val dispatcher: Dispatch
         }
         ongoingRequests.removeAll(event.orderIds.toSet())
     }
+
+    @Suppress("unused", "ForbiddenComment")
+    @Subscribe(threadMode = ThreadMode.BACKGROUND)
+    fun onOrderListFetched(event: OnListChanged) {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/FetchOrdersRepository.kt:73]: Updates the last update every time the first page is fetched

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/ObserveOrdersListLastUpdate.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.woocommerce.android.ui.orders.list
+
+import com.woocommerce.android.background.LastUpdateDataStore
+import javax.inject.Inject
+
+class ObserveOrdersListLastUpdate @Inject constructor(private val lastUpdateDataStore: LastUpdateDataStore) {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/ObserveOrdersListLastUpdate.kt:6]: Use case to observe the last update of the order list. I did not add unit tests to this class due to its simplicity.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/ShouldUpdateOrdersList.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.woocommerce.android.ui.orders.list
+
+import com.woocommerce.android.background.LastUpdateDataStore
+import kotlinx.coroutines.flow.first
+import javax.inject.Inject
+
+class ShouldUpdateOrdersList @Inject constructor(private val lastUpdateDataStore: LastUpdateDataStore) {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/ShouldUpdateOrdersList.kt:7]: Use case to check if we should refresh the orders list cache. I did not add unit tests to this class due to its simplicity.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/StoreOrdersListLastUpdate.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.woocommerce.android.ui.orders.list
+
+import com.woocommerce.android.background.LastUpdateDataStore
+import javax.inject.Inject
+
+class StoreOrdersListLastUpdate @Inject constructor(private val lastUpdateDataStore: LastUpdateDataStore) {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/StoreOrdersListLastUpdate.kt:6]: Use case to persist the last update timestamp for the orders list. I did not add unit tests to this class due to its simplicity.

**Review by ThomazFB — APPROVED** [review]
Great work here, @atorresveiga, really nice to see the `last update` logic introduced in the Analytics Hub being expanded to other sections of the app. Code looks code and feature works as described.

My only point that I would like to discuss is  that the three Use cases you defined and didn't add unit test to them due to their simplicity are suggesting a `manager` class wrapping the new data store.  

You have three use cases working on top of the `LastUpdateDataStore` class:
- `ObserveOrdersListLastUpdate`
- `ShouldUpdateOrdersList`
- `StoreOrdersListLastUpdate`

I understand that they shouldn't be added to the `LastUpdateDataStore` as that data store definition is agnostic of the Order context, but those three use cases have one thing in common: they combine the context of Orders + LastUpdateDataStore. One reading, one writing, and the other asking to that context. 

I think that a `OrdersLastUpdateManager` class could do those three things, instead of scattering this logic into multiple use cases. This gets even more highlighted due to those three use cases only having one single dependency, which is the data store itself. I usually prefer to resort to use cases when I have to combine multiple dependencies into a single complex logic unit.

This is just a suggestion, though, it doesn't block the PR in any sense.  Let me know what you think!

**Comment by atorresveiga** [comment]
Thanks for the suggestion @ThomazFB! I appreciate you looking out for code reusability and maintainability.

> I think that a OrdersLastUpdateManager class could do those three things, instead of scattering this logic into multiple use cases. This gets even more highlighted due to those three use cases only having one single dependency, which is the data store itself. I usually prefer to resort to use cases when I have to combine multiple dependencies into a single complex logic unit.

Here's my perspective on the use cases and the potential OrdersLastUpdateManager:

Benefits of Use Cases (Current Approach):

- Clarity and Focus: Separate use cases clearly define specific functionalities related to order updates. This improves code readability and understanding.
- Testability: Each use case can be easily unit tested in isolation, ensuring individual functionalities work as expected.
- Gradual Introduction: Use cases allow for introducing functionalities incrementally, making the code more approachable for future modifications.

Potential OrdersLastUpdateManager:

- Centralized Management: A manager class could offer a central point for order update logic, potentially improving organization.
- Reduced Code Duplication: Combining logic into a manager might reduce code duplication if functionalities become more complex in the future.

Considering that use cases are a common pattern in the code base, the isolation benefits they bring, and the simplicity of these specific use cases, I'm leaning toward maintaining these simple use cases for now. If the logic or dependencies grow significantly in the future, then an `OrdersLastUpdateManager` could be an excellent option to explore.

Nevertheless, your comment made me realize I was missing a significant opportunity to take advantage of one of the crucial benefits of use cases (Testability). I will implement the unit tests for these use cases.

I'm open to further discussion and finding the best approach for maintainable and reusable code. 😄

**Pull request merged**